openmls/src/treesync/diff.rs

```diff
@@ -0,0 +1,806 @@
+                // until it's non-blank or a leaf.
+                while self.diff.node(right_child_id)?.node().is_none()
+                    && !self.diff.is_leaf(right_child_id)
+                {
+                    right_child_id = self.diff.left_child(right_child_id)?;
+                }
+                // If the "right child" is a non-blank node, we continue,
+                // otherwise it has to be a blank leaf node and the check
+                // fails.
+                if let Some(right_child) = self.diff.node(right_child_id)?.node() {
+                    // Perform the check with the parent hash of the "right
+                    // child" and the left child resolution.
+                    let mut left_child_resolution =
+                        self.resolution(left_child_id, &HashSet::new())?;
+                    // Filter unmerged leaves from resolution.
+                    self.filter_resolution(parent_node, &mut left_child_resolution)?;
+                    let node_hash = parent_node.compute_parent_hash(
+                        backend,
+                        ciphersuite,
+                        parent_node.parent_hash(),
+                        &left_child_resolution,
+                    )?;
+                    if let Some(right_child_parent_hash) = right_child.parent_hash()? {
+                        if node_hash == right_child_parent_hash {
+                            // If the hashes match, we continue with the next node.
+                            continue;
+                        };
+                    }
+                    // If the hash doesn't match, or the leaf doesn't have a
+                    // parent hash extension (the `None` case in the `if let`
+                    // above), the verification fails.
+                }
+                return Err(TreeSyncDiffError::InvalidParentHash);
+            } else {
+                continue;
+            }
+        }
+        Ok(())
+    }
+
+    /// This turns the diff into a staged diff. In the process, the diff
+    /// computes and sets the new tree hash.
+    pub(crate) fn into_staged_diff(
+        mut self,
+        backend: &impl OpenMlsCryptoProvider,
+        ciphersuite: &Ciphersuite,
+    ) -> Result<StagedTreeSyncDiff, TreeSyncDiffError> {
+        let new_tree_hash = self.compute_tree_hash(backend, ciphersuite)?;
+        #[cfg(test)]
+        self.verify_parent_hashes(backend, ciphersuite)
+            .expect("error verifying parent hashes");
+        Ok(StagedTreeSyncDiff {
+            diff: self.diff.into(),
+            new_tree_hash,
+        })
+    }
+
+    /// Helper function to compute and set the tree hash of the given node and
+    /// all nodes below it in the tree. This function respects cached tree hash
+    /// values. If a cached value is found it is returned without further
+    /// computation of hashes of the node or the nodes below it.
+    fn set_tree_hash(
```

Comment:
Something like `compute_tree_hash`? `set_` doesn't sound like computing and the cashing is confusing then.